app/controllers/posts_controller.rb

```diff
@@ -48,9 +49,7 @@ def unread
 
   def mark
     posts = Post.where(id: params[:marked_ids])
-    posts = posts.select do |post|
-      post.visible_to?(current_user)
-    end
+    posts = posts.select { |post| post.visible_to?(current_user) }
```

Comment:
Purely informationally: `posts` here is indeed a Queryset not an Array so the issue with `select!` was probably Querysets doing nonsense to differentiate from the SQL select function.